cmd/soroban-rpc/lib/preflight/src/lib.rs

```diff
@@ -0,0 +1,232 @@
+extern crate libc;
+extern crate soroban_env_host;
+
+use soroban_env_host::budget::Budget;
+use soroban_env_host::storage::{self, AccessType, SnapshotSource, Storage};
+use soroban_env_host::xdr::{
+    self, AccountId, HostFunction, LedgerEntry, LedgerKey, ReadXdr, ScHostStorageErrorCode,
+    WriteXdr,
+};
+use soroban_env_host::{Host, HostError, LedgerInfo};
+use std::convert::TryInto;
+use std::ffi::{CStr, CString};
+use std::panic;
+use std::ptr::null_mut;
+use std::rc::Rc;
+use xdr::LedgerFootprint;
+
+// TODO: we may want to pass callbacks instead of using global functions
+extern "C" {
+    // LedgerKey XDR in base64 string to LedgerEntry XDR in base64 string
+    fn SnapshotSourceGet(
+        handle: libc::uintptr_t,
+        ledger_key: *const libc::c_char,
+    ) -> *const libc::c_char;
+    // LedgerKey XDR in base64 string to bool
+    fn SnapshotSourceHas(handle: libc::uintptr_t, ledger_key: *const libc::c_char) -> libc::c_int;
+    // Free Strings returned from Go functions
+    fn FreeGoCString(str: *const libc::c_char);
+}
+
+struct CSnapshotSource {
+    handle: libc::uintptr_t,
+}
+
+impl SnapshotSource for CSnapshotSource {
+    fn get(&self, key: &LedgerKey) -> Result<LedgerEntry, HostError> {
+        let key_xdr = key.to_xdr_base64().unwrap();
+        let key_cstr = CString::new(key_xdr).unwrap();
+        let res = unsafe { SnapshotSourceGet(self.handle, key_cstr.as_ptr()) };
+        if res.is_null() {
+            return Err(HostError::from(
+                ScHostStorageErrorCode::AccessToUnknownEntry,
+            ));
+        }
+        let res_cstr = unsafe { CStr::from_ptr(res) };
+        let res_str = res_cstr.to_str().unwrap();
+        // TODO: use a proper error
+        let entry = LedgerEntry::from_xdr_base64(res_str)
+            .map_err(|_| ScHostStorageErrorCode::UnknownError)?;
+        unsafe { FreeGoCString(res) };
+        Ok(entry)
+    }
+
+    fn has(&self, key: &LedgerKey) -> Result<bool, HostError> {
+        let key_xdr = key.to_xdr_base64().unwrap();
+        let key_cstr = CString::new(key_xdr).unwrap();
```

Comment:
`unwrap` vs `?` here?

Reply:
❗ +1 It would be ideal if we could eliminate all the `.unwrap()` calls.

If we remove the base64 back and forth we should reduce some of the error cases needed to handle, since we wouldn't need to dance around Rust's unicode-only string types.

Reply:
> If we remove the base64 back and forth we should reduce some of the error cases needed to handle

That would require:
1. creating (and book-keeping) C stubs for ledger entries 
2. Converting the C entries version back a forth Go<->C<->Rust

I would very much like to avoid that

Reply:
> * creating (and book-keeping) XDR C stubs for ledger entries

I'm not sure what this means. I wasn't suggesting passing C structs back and forth, but rather just the XDR binary, instead of XDR binary that is also base64 encoded. I believe it would mean not needing to use `CString`. But yes there would still be failure cases for `.to_xdr()` instead of `.to_xdr_base64()`.

Regardless, we should be able to return a `HostError` in each of these error cases instead of panicking. The `HostError` can be a little awkward to work with, but we can create simple ones with the `Sc` error codes, which I see us doing in some lines above (e.g. `.map_err(|_| ScHostStorageErrorCode::UnknownError)?`).

We can ask @graydon @jayz22 about how to attach more information to the host error too.

Reply:
> > * creating (and book-keeping) XDR C stubs for ledger entries
> 
> I'm not sure what this means. I wasn't suggesting passing C structs back and forth, but rather just the XDR binary, instead of XDR binary that is also base64 encoded. I believe it would mean not needing to use `CString`. But yes there would still be failure cases for `.to_xdr()` instead of `.to_xdr_base64()`.

1. The `preflight_host_function()` caller needs base64-encoded results
2. If we pass binary-buffers through C, we need to pass both the buffer and the length separately, which is a pain, particularly for return values (e.g. `extern char *SnapshotSourceGet(uintptr_t handle, char *ledger_key);`)

So, for simplicitly, I think it's better to just use base64 everywhere.

> Regardless, we should be able to return a `HostError` in each of these error cases instead of panicking. The `HostError` can be a little awkward to work with, but we can create simple ones with the `Sc` error codes, which I see us doing in some lines above (e.g. `.map_err(|_| ScHostStorageErrorCode::UnknownError)?`).
> 
> We can ask @graydon @jayz22 about how to attach more information to the host error too.

Happy to use `HostError` but I didn't see how to include generic information (and the existing errors wouldn't convey what happened). I will ask for help.

> Or is there no consequence of a panic in this code? I'm not sure how a panic here will impact the integrity or the Go program, or the impact to unfreed memory.


The Go program should be fine, since the panic shouldn't propagate there (we catch the panic explicitly).

As for the memory, I think that we can only leak memory by not calling `FreeGoCString()`, which could happen if `res_cstr.to_str().unwrap();` panics. I will try to return a `HostError` there.

Reply:
Part of this is tracked at https://github.com/stellar/soroban-tools/issues/367